### ISSUE TITLE
Create diagnostic and code fix for mismatched argument constraint types

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#1163]
From https://github.com/FakeItEasy/FakeItEasy/pull/1156#discussion_r125862320

Most type mismatches will be detected by the compiler, but if the types can be coerced, the compiler lets us down.

@blairconrad:
> Maybe we should have a diagnostic for the case where the type of the constraint is mismatched!

@thomasleveque:
> Oh, I had never thought of that. Currently `A<int>._` won't match if the actual parameter type is `long`. Yes, we should totally add a diagnostic for this.

An alternative would be to make the argument constraints more forgiving, but that seems like a tonne of work.

**Comment by thomaslevesque** [comment]
To clarify, here's an example:

```csharp
public interface IFoo
{
    string Bar(long x);
}

...

var foo = A.Fake<IFoo>();
A.CallTo(() => foo.Bar(A<int>._)).Returns("hello");
```
It compiles, because `int` is implicitly converted to `long`, but no call will ever match the constraint because it's of the wrong type.

**Comment by blairconrad** [comment]
Self-assigning. I might as well get in on this analyzer action.

**Comment by thomaslevesque** [comment]
> Self-assigning. I might as well get in on this analyzer action.

👍 

Rather than creating a completely new analyzer, we should probably consider adding the new diagnostic to the existing `ArgumentConstraintNullabilityMismatchAnalyzer`, since most of the code will be identical (a single analyzer can produce multiple diagnostics). Of course, it will have to be merged first... I'll make the necessary changes now so you can merge it.

**Comment by blairconrad** [comment]
There's no rush at all. I already have a small proof of concept built off your branch.
Also, I may not jump onto the meat of the issue right away. I'm being greedy by taking it, but there's no particular rush for this. No big deal if it waits until after 4.0, IMO.

**Comment by thomaslevesque** [comment]
> No big deal if it waits until after 4.0, IMO.

Sure, it's not a breaking change anyway (well, it might cause warnings on existing broken code, but that doesn't really count as breaking)

**Comment by blairconrad** [comment]
I'd be inclined to make the diagnostic an error.

**Comment by thomaslevesque** [comment]
> I'd be inclined to make the diagnostic an error.

Yes, that makes sense.

**Comment by thomaslevesque** [comment]
Closed by #1192

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 4.1.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/4.1.0).
